packages/components/src/DatePicker/DatePicker.tsx

```diff
@@ -115,7 +121,74 @@ export function DatePicker({
   const datePickerClassNames = classnames(styles.datePicker, {
     [styles.inline]: inline,
   });
-  const { pickerRef } = useEscapeKeyToCloseDatePicker(open, ref);
+  const pickerRef = useRef<ReactDatePicker>(null);
+  const pickerId = useRef(useId());
+
+  /**
+   * The onChange callback on ReactDatePicker returns a Date and an Event, but
+   * the onChange in our interface only provides the Date. Simplifying the code
+   * by removing this function and passing it directly to the underlying
+   * component breaks tests both here and downstream (i.e. the pattern
+   * `expect(onChange).toHaveBeenCalledWith(date)` is commonly used and would
+   * fail).
+   */
+  function handleChange(value: Date /* , event: React.SyntheticEvent */) {
+    onChange(value);
+  }
+
+  function handleCalendarOpen() {
+    document.dispatchEvent(
+      new CustomEvent(DATEPICKER_SWITCH_EVENT, {
+        detail: { id: pickerId.current },
+      }),
+    );
+
+    pickerRef.current?.setOpen(true);
+    setOpen(true);
+  }
+
+  function handleCalendarClose() {
+    pickerRef.current?.setOpen(false);
+    setOpen(false);
+  }
+
+  // Listen for close events from other DatePickers
+  useEffect(() => {
+    const handleClose = (event: CustomEvent<DatePickerCloseDetails>) => {
+      if (event.detail.id !== pickerId.current) {
+        pickerRef.current?.setOpen(false);
+        setOpen(false);
+      }
+    };
+
+    document.addEventListener(
+      DATEPICKER_SWITCH_EVENT,
+      handleClose as EventListener,
+    );
+
+    return () => {
+      document.removeEventListener(
+        DATEPICKER_SWITCH_EVENT,
+        handleClose as EventListener,
+      );
+    };
+  }, []);
+
+  // Handle ESC key
+  useEffect(() => {
+    const escFunction = (event: KeyboardEvent) => {
+      if (event.key === "Escape" && open) {
+        pickerRef.current?.setOpen(false);
+        event.stopPropagation();
+      }
+    };
+
+    document.addEventListener("keydown", escFunction, true);
+
+    return () => {
+      document.removeEventListener("keydown", escFunction, true);
+    };
+  }, [open]);
```

Comment:
Some of the above is not new, like Github is making it look, just moved before the return